sympy/stats/stochastic_process_types.py

```diff
@@ -520,20 +521,53 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
         new_given_condition=self.replace_with_index(new_given_condition)
 
         if isinstance(condition, Relational):
-            rv, states = (list(condition.atoms(RandomIndexedSymbol))[0], condition.as_set())
             if isinstance(new_given_condition, And):
                 gcs = new_given_condition.args
             else:
                 gcs = (new_given_condition, )
-            grvs = new_given_condition.atoms(RandomIndexedSymbol)
-
-            min_key_rv = None
-            for grv in grvs:
-                if grv.key <= rv.key:
-                    min_key_rv = grv
-            if min_key_rv == None:
+            min_key_rv = list(new_given_condition.atoms(RandomIndexedSymbol))
+            rv = list(condition.atoms(RandomIndexedSymbol))
+
+            if len(min_key_rv):
+                min_key_rv = min_key_rv[0]
+                for r in rv:
+                    if min_key_rv.key > r.key:
+                        return Probability(condition)
+            else:
+                min_key_rv = None
                 return Probability(condition)
 
+            if len(rv) > 1:
+                rv = rv[:2]
+                if rv[0].key < rv[1].key:
+                        rv[0], rv[1] = rv[1], rv[0]
+                s = Float(0)
+                n = len(self.state_space)
+
+                if isinstance(condition, Eq) or isinstance(condition, Ne):
+                    for i in range(0, n):
+                        s += P(Eq(rv[0], i), Eq(rv[1], i)) * P(Eq(rv[1], i), new_given_condition)
+                    return s if isinstance(condition, Eq) else 1-s
+                else:
+                    upper = 0
+                    greater = False
+                    if isinstance(condition, Ge) or isinstance(condition, Lt):
+                        upper = 1
+                    if isinstance(condition, Gt) or isinstance(condition, Ge):
+                        greater = True
+
+                    for i in range(0, n):
+                        if i <= n//2:
+                            for j in range(0, i + upper):
+                                s += P(Eq(rv[0], i), Eq(rv[1], j)) * P(Eq(rv[1], j), new_given_condition)
+                        else:
+                            s+=P(Eq(rv[0], i), new_given_condition)
```

Comment:
```suggestion
                            s += self.probability(Eq(rv[0], i), new_given_condition)
```
Calling `P` again is redundant when we know the job will be actually done by `self.probability`.

Reply:
Yes, sorry about that